sdks/java/core/src/main/java/org/apache/beam/sdk/runners/AppliedPTransform.java

```diff
@@ -70,4 +70,19 @@
   public abstract TransformT getTransform();
 
   public abstract Pipeline getPipeline();
+
+  @Override
+  public boolean equals(Object other) {
```

Comment:
I'm going to say the same thing you said to me - probably the map should be keyed by the stable name or an equivalence wrapper, so the object itself can have context-unaware equality.